### PR TITLE
Feat/collect keys sorted

### PR DESCRIPTION
**Pull request by haraldLmueller**

### CODE REVIEW

**Comment by alexbrdn** [comment]
I have to question the utility of this. For one, there's no way of getting the same for secondary indexes. Secondly, I don't know where would it be useful, except in tests.

**Comment by haraldLmueller** [comment]
In my case it is useful to create a hash over all elements in the list. Without an stable order, it would create a different hash on each try.  It also enables you to create a paginated list.

File: map.go [hunk 1]
```diff
@@ -1,21 +1,27 @@
...
+
+	"golang.org/x/exp/constraints"
+)
 
 // IndexMap is a map supports seeking data with more indexes.
 // Serializing a IndexMap as JSON results in the same as serializing a map,
 // the result doesn't contain the index information, only data.
 // NOTE: DO NOT insert nil value into the IndexMap
-type IndexMap[K comparable, V any] struct {
+type IndexMap[K constraints.Ordered, V any] struct {
```
Comment (alexbrdn) [inline map.go:14]: IMO, this is too narrow of a constraint (no pun intended) on possible key types. When using `Ordered`, arrays and comparable structs would not be accepted as keys. IndexMap is supposed to be like a map, but with additional features, so it should accept all key types that a normal map can have. Maybe you should instead make a wrapper around IndexMap for whatever you're trying to achieve.

File: map.go [hunk 2]
```diff
@@ -306,6 +325,26 @@ func (imap *IndexMap[K, V]) CollectBy(indexName string) ([]any, [][]*V) {
...
+		for k := range imap.primaryIndex.inner {
+			keys = append(keys, k)
+		}
+		sort.SliceStable(keys, func(i, j int) bool {
+			return keys[i] < keys[j]
+		})
+		imap.sorted = &keys
+	}
+	// return a copy to avoid that the original was modified
+	ret := *imap.sorted
```
Comment (alexbrdn) [inline map.go:344]: this is not a copy. proof: https://go.dev/play/p/cTWxvaUbsTN

**Review by alexbrdn — COMMENTED** [review]
Respectfully, please reconsider narrowing of the possible keys space

**Pull request closed without merging**